src/python/pants/backend/go/goals/package_binary_integration_test.py

```diff
@@ -106,6 +106,59 @@ def test_package_simple(rule_runner: RuleRunner) -> None:
     assert result.stdout == b"Hello world!\n"
 
 
+def test_package_third_party_requires_main(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "go.mod": dedent(
+                """\
+                module foo.example.com
+                go 1.17
+
+                require github.com/tgolsson/example-pants-third-party v0.0.0-20221101220057-1a7167a87ec5 // indirect
+                """
+            ),
+            "BUILD": dedent(
+                """\
+                go_mod(name='mod')
+                go_binary(name="bin", main='//:mod#github.com/tgolsson/example-pants-third-party/pkg/hello')
+                """
+            ),
+        }
+    )
+    binary_tgt = rule_runner.get_target(Address("", target_name="bin"))
+    with engine_error(ValueError, contains="but uses package name `hello` instead of `main`"):
+        build_package(rule_runner, binary_tgt)
+
+
+def test_package_third_party_can_run(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "go.mod": dedent(
+                """\
+                module foo.example.com
+                go 1.17
+
+                require github.com/tgolsson/example-pants-third-party v0.0.0-20221101220057-1a7167a87ec5 // indirect
+                """
+            ),
+            "BUILD": dedent(
+                """\
+                go_mod(name='mod')
+                go_binary(name="bin", main='//:mod#github.com/tgolsson/example-pants-third-party/cmd/hello')
```

Comment:
Maybe move this third party dependency into the Pants repo for the test? See `test_third_party_package_embed` in `src/python/pants/backend/go/util_rules/embed_integration_test.py` to see how to do this. My concern here is having an external dependency for the test on a non-pantsbuild org repo.

Reply:
And it will still be a "third party" dependency for Pants and `go`. Basically that other test simulates a Go module proxy site but within the test.

Reply:
Will do! 

While you're here; do you have any ideas about the test failures? I cannot reproduce them locally.

```
ValueError: For `go_mod` target `//:mod`, the go.sum file is incomplete because it was updated while processing third-party dependency `github.com/tgolsson/example-pants-third-party`. Please re-generate the go.sum file by running `go mod download all` in the module directory. (Pants does not currently have support for updating the go.sum checksum database itself.)
```

Reply:
This is happening due to my bug fix in https://github.com/pantsbuild/pants/pull/17811. This test doesn't have a `go.sum` but needs to have one since it is using a third party module. Previously, `go` was silently updating the `go.sum` copied into the execution sandbox, but Pants never detected that fact and ignored it which is not good for software supply chain security.

The other test includes code to compute the correct checksm for put into a go.sum file (which it also does). So this test failure should be fixed as a side effect of switching over to simulating a Go module proxy in the test.

Reply:
Done and done, thank you!